File: src/binance_mcp/tools/account.py
"""Account management MCP tools for Binance."""

from typing import Any
from mcp.types import Tool
from binance.client import Client


def get_account_tools() -> list[Tool]:
    """Get list of account management tools."""
    return [
        Tool(
            name="get_account_info",
            description="Get current account information including balances",
            inputSchema={
                "type": "object",
                "properties": {},
                "required": []
            }
        ),
        Tool(
            name="get_asset_balance",
            description="Get balance for a specific asset",
            inputSchema={
                "type": "object",
                "properties": {
                    "asset": {
                        "type": "string",
                        "description": "Asset symbol (e.g., 'BTC', 'USDT')"
                    }
                },
                "required": ["asset"]
            }
        ),
        Tool(
            name="get_account_trades",
            description="Get trades for a specific symbol",
            inputSchema={
                "type": "object",
                "properties": {
                    "symbol": {
                        "type": "string",
                        "description": "Trading pair symbol (e.g., 'BTCUSDT')"
                    },
                    "limit": {
                        "type": "number",
                        "description": "Number of trades to retrieve (default: 500, max: 1000)",
                        "default": 500
                    },
                    "from_id": {
                        "type": "number",
                        "description": "Trade ID to fetch from"
                    }
                },
                "required": ["symbol"]
            }
        ),
        Tool(
            name="get_account_status",
            description="Get account status (normal, margin, futures)",
            inputSchema={
                "type": "object",
                "properties": {},
                "required": []
            }
        ),
        Tool(
            name="get_trade_fee",
            description="Get trade fee for a symbol",
            inputSchema={
                "type": "object",
                "properties": {
                    "symbol": {
                        "type": "string",
                        "description": "Trading pair symbol (e.g., 'BTCUSDT')"
                    }
                },
                "required": []
            }
        ),
        Tool(
            name="get_asset_dividend_history",
            description="Get asset dividend history",
            inputSchema={
                "type": "object",
                "properties": {
                    "asset": {
                        "type": "string",
                        "description": "Asset symbol (e.g., 'BTC')"
                    },
                    "start_time": {
                        "type": "number",
                        "description": "Start time in milliseconds"
                    },
                    "end_time": {
                        "type": "number",
                        "description": "End time in milliseconds"
                    }
                },
                "required": []
            }
        ),
    ]


def handle_account_tool(client: Client, name: str, arguments: Any) -> Any:
    """Handle account tool execution."""
    if name == "get_account_info":
        return client.get_account()

    elif name == "get_asset_balance":
        return client.get_asset_balance(asset=arguments["asset"])

    elif name == "get_account_trades":
        return client.get_my_trades(
            symbol=arguments["symbol"],
            limit=arguments.get("limit", 500),
            fromId=arguments.get("from_id")
        )

    elif name == "get_account_status":
        return client.get_account_status()

    elif name == "get_trade_fee":
        symbol = arguments.get("symbol")
        if symbol:
            return client.get_trade_fee(symbol=symbol)
        return client.get_trade_fee()

    elif name == "get_asset_dividend_history":
        return client.get_asset_dividend_history(
            asset=arguments.get("asset"),
            startTime=arguments.get("start_time"),
            endTime=arguments.get("end_time")
        )

    return None

File: src/tools/account.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { BinanceClient } from "../client.js";
import { success, error } from "../utils.js";

export function registerAccountTools(server: McpServer, client: BinanceClient) {
  server.tool(
    "get_account_info",
    "Get current account information including all asset balances",
    {},
    async () => {
      try {
        return success(await client.getAccountInfo());
      } catch (err) {
        return error(err);
      }
    },
  );

  server.tool(
    "get_asset_balance",
    "Get balance for a specific asset",
    {
      asset: z
        .string()
        .describe("Asset symbol (e.g. BTC, USDT, ETH)"),
    },
    async ({ asset }) => {
      try {
        return success(await client.getAssetBalance(asset));
      } catch (err) {
        return error(err);
      }
    },
  );

  server.tool(
    "get_account_trades",
    "Get trade history for a specific trading pair",
    {
      symbol: z.string().describe("Trading pair (e.g. BTCUSDT)"),
      limit: z
        .number()
        .optional()
        .describe("Number of trades (default: 500, max: 1000)"),
      from_id: z
        .number()
        .optional()
        .describe("Trade ID to fetch from"),
    },
    async ({ symbol, limit, from_id }) => {
      try {
        return success(await client.getAccountTrades(symbol, limit, from_id));
      } catch (err) {
        return error(err);
      }
    },
  );

  server.tool(
    "get_account_status",
    "Get account status (normal or restricted)",
    {},
    async () => {
      try {
        return success(await client.getAccountStatus());
      } catch (err) {
        return error(err);
      }
    },
  );

  server.tool(
    "get_trade_fee",
    "Get trading fee rates for a symbol or all symbols",
    {
      symbol: z
        .string()
        .optional()
        .describe("Trading pair (e.g. BTCUSDT)"),
    },
    async ({ symbol }) => {
      try {
        return success(await client.getTradeFee(symbol));
      } catch (err) {
        return error(err);
      }
    },
  );

  server.tool(
    "get_asset_dividend_history",
    "Get asset dividend/distribution history",
    {
      asset: z.string().optional().describe("Asset symbol (e.g. BTC)"),
      start_time: z
        .number()
        .optional()
        .describe("Start time in milliseconds"),
      end_time: z.number().optional().describe("End time in milliseconds"),
    },
    async ({ asset, start_time, end_time }) => {
      try {
        return success(
          await client.getAssetDividendHistory(asset, start_time, end_time),
        );
      } catch (err) {
        return error(err);
      }
    },
  );

  server.tool(
    "get_dust_assets",
    "Get list of assets that can be converted to BNB (dust balances below 0.0012 BTC)",
    {},
    async () => {
      try {
        return success(await client.getDustAssets());
      } catch (err) {
        return error(err);
      }
    },
  );

  server.tool(
    "convert_dust_to_bnb",
    "Convert small balances (dust) to BNB. 2% fee applies. Can be used once every 6 hours.",
    {
      assets: z
        .array(z.string())
        .describe("List of asset symbols to convert (e.g. ['ETH', 'XRP', 'SOL'])"),
    },
    async ({ assets }) => {
      try {
        return success(await client.convertDustToBnb(assets));
      } catch (err) {
        return error(err);
      }
    },
  );
}
